cirq/experiments/grid_parallel_two_qubit_xeb.py

```diff
@@ -0,0 +1,510 @@
+    metadata = load(metadata_params, base_dir=base_dir)
+    qubits = metadata['qubits']
+    num_circuits = metadata['num_circuits']
+    repetitions = metadata['repetitions']
+    cycles = metadata['cycles']
+    layers = metadata['layers']
+
+    coupled_qubit_pairs = _coupled_qubit_pairs(qubits)
+    all_active_qubit_pairs = []
+    xeb_results = {}  # type: Dict[GridQubitPair, CrossEntropyResult]
+
+    with multiprocessing.Manager() as manager:
+        # Each combination of qubit pair and parent circuit can be handled in a
+        # separate process
+        # Reason for type: ignore: https://github.com/python/mypy/issues/4678
+        numerators = manager.dict()  # type: ignore
+        denominators = manager.dict()  # type: ignore
+        arguments = []
+        # Construct arguments to be mapped by multiprocessing
+        for layer in layers:
+            active_qubit_pairs = [
+                pair for pair in coupled_qubit_pairs if pair in layer
+            ]
+            all_active_qubit_pairs.extend(active_qubit_pairs)
+            for i in range(num_circuits):
+                circuit_params = GridParallelXEBCircuitParameters(
+                    data_collection_id=data_collection_id,
+                    layer=layer,
+                    circuit_index=i)
+                circuit = load(circuit_params, base_dir=base_dir)
+                trial_results = []
+                for depth in cycles:
+                    trial_result_params = GridParallelXEBTrialResultParameters(
+                        data_collection_id=data_collection_id,
+                        layer=layer,
+                        depth=depth,
+                        circuit_index=i)
+                    trial_result = load(trial_result_params, base_dir=base_dir)
+                    trial_results.append(trial_result)
+                for qubit_pair in active_qubit_pairs:
+                    arguments.append((qubit_pair, qubits, circuit, cycles,
+                                      trial_results, numerators, denominators))
+
+        # Initialize lists to store numerators and denominators.
+        for qubit_pair, depth in itertools.product(all_active_qubit_pairs,
+                                                   cycles):
+            # Reason for type: ignore:
+            # https://github.com/python/mypy/issues/4678
+            numerators[(qubit_pair, depth)] = manager.list()  # type: ignore
+            denominators[(qubit_pair, depth)] = manager.list()  # type: ignore
+
+        # Compute the numerators and denominators
+        num_processors = min(num_processors, len(arguments))
+        with multiprocessing.Pool(num_processors) as pool:
+            pool.starmap(_get_fidelity_estimator_components, arguments)
+
+        # Calculate the results
+        for qubit_pair in all_active_qubit_pairs:
+            data = []
+            for depth in cycles:
+                fidelity = (sum(numerators[(qubit_pair, depth)]) /
```

Comment:
For estimating fidelity from XEB data you should use one of the estimators in [fidelity_estimation.py](https://github.com/quantumlib/Cirq/blob/master/cirq/experiments/fidelity_estimation.py). Alternatively, if none of them fit your requirements, you should add your own there and call it from here.

Ideally, one should be able to use the code with any fidelity estimator they choose(*).

We should avoid having disjoint, incompatible sets of machinery for performing related tasks.

---
(*) There are good reasons to prefer different estimators in different circumstances, e.g. estimators generally have different variance.

Reply:
I've defined a new fidelity estimator function, but I don't see a way to make the API compatible. The reason is that the estimator I'm using uses additional information to the measured bitstring probabilities; it also uses the probabilities for all possible bitstrings. Thus, this estimator is only practical for a small Hilbert space dimension.